File: client.java/src/main/java/dev/resolvedor/client/backup/BackupClient.java
package dev.resolvedor.client.backup;

import dev.resolvedor.ServiceGenerator;
import java.io.File;
import okhttp3.MediaType;
import okhttp3.MultipartBody;
import okhttp3.RequestBody;
import okhttp3.ResponseBody;
import retrofit2.Call;
import retrofit2.Response;

public class BackupClient {

    private static final MediaType DEFAULT_FILE_MEDIA_TYPE = MediaType.get("application/octet-stream");

    private final ServiceGenerator generator;
    private final String tokenSecret;

    public BackupClient(String urlBackupServer, String tokenSecret) {
        this.generator = new ServiceGenerator(urlBackupServer);
        this.tokenSecret = tokenSecret;
    }

    public void backup(String sha2Code, File file) {
        BackupService service = generator.createService(BackupService.class, tokenSecret);

        RequestBody sha2CodeBody = RequestBody.create(MediaType.parse(DEFAULT_FILE_MEDIA_TYPE.toString()), sha2Code);
        RequestBody fileBody = RequestBody.create(file, DEFAULT_FILE_MEDIA_TYPE);
        MultipartBody.Part filePart = MultipartBody.Part.createFormData("file_data", file.getName(), fileBody);

        Call<ResponseBody> callSync = service.backup(sha2CodeBody, filePart);

        try {
            Response<ResponseBody> response = callSync.execute();
            if (response.isSuccessful()) {
                ResponseBody body = response.body();
                System.out.println(body != null ? body.string() : "");
            } else {
                String errorBody = response.errorBody() != null ? response.errorBody().string() : "";
                System.out.println("Request failed with code: " + response.code());
                if (!errorBody.isEmpty()) {
                    System.out.println(errorBody);
                }
            }
        } catch (Exception e) {
            System.out.println(e.getMessage());
        }
    }
}

File: client.kotlin/app/src/main/kotlin/dev/resolvedor/client/backup/BackupClient.kt
package dev.resolvedor.client.backup

import dev.resolvedor.client.ServiceGenerator
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.MultipartBody
import okhttp3.RequestBody.Companion.asRequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import retrofit2.Call
import java.io.File

class BackupClient(urlBackupServer: String) {

    private val generator: ServiceGenerator = ServiceGenerator(urlBackupServer)

    fun backup(file: File, sha2: String, tokenSecret: String) {
        val service = generator.createService(BackupService::class.java, tokenSecret)

        val sha2Body = sha2.toRequestBody("text/plain".toMediaTypeOrNull())
        val fileRequestBody = file.asRequestBody("application/octet-stream".toMediaTypeOrNull())
        val filePart = MultipartBody.Part.createFormData("file_data", file.name, fileRequestBody)

        try {
            val call: Call<BackupResponse> = service.postBackup(sha2Body, filePart)
            val response = call.execute()
            if (response.isSuccessful) {
                println(response.body()?.message)
            } else {
                println("Request failed with code: ${response.code()}")
            }
        } catch (e: Exception) {
            println(e.message)
        }
    }
}

